src/sage/matroids/circuits_matroid.pyx

```diff
@@ -0,0 +1,613 @@
+            True
+            sage: M.reset_name()
+            sage: loads(dumps(M))
+            Matroid of rank 5 on 10 elements with 45 circuits
+        """
+        import sage.matroids.unpickling
+        data = (self._groundset, frozenset(self._C), self.get_custom_name())
+        version = 0
+        return sage.matroids.unpickling.unpickle_circuits_matroid, (version, data)
+
+    # enumeration
+
+    cpdef bases(self) noexcept:
+        r"""
+        Return the bases of the matroid.
+
+        OUTPUT: a SetSystem
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+        """
+        cdef SetSystem B
+        B = SetSystem(list(self.groundset()))
+        from itertools import combinations
+        for S in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(S):
+                    flag = False
+                    break
+            if flag:
+                B.append(S)
+        return B
+
+    def bases_iterator(self):
+        r"""
+        Return an iterator over the bases of the matroid.
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+        """
+        from itertools import combinations
+        for B in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(B):
+                    flag = False
+                    break
+            if flag:
+                yield frozenset(B)
+
+    cpdef circuits(self, k=None) noexcept:
+        """
+        Return the circuits of the matroid.
+
+        OUTPUT: a SetSystem
+        """
+        cdef SetSystem C
+        C = SetSystem(list(self.groundset()))
+        if k:
+            for c in self._k_C[k]:
+                C.append(c)
+        else:
+            for i in self._k_C:
+                for c in self._k_C[i]:
+                    C.append(c)
+        return C
+
+    def circuits_iterator(self, k=None):
+        """
+        Return an iterator over the circuits of the matroid.
+        """
+        if k:
+            for C in self._k_C[k]:
+                yield C
+        else:
+            for i in self._k_C:
+                for C in self._k_C[i]:
+                    yield C
+
+    cpdef nonspanning_circuits(self) noexcept:
+        """
+        Return the list of nonspanning circuits of the matroid.
+
+        OUTPUT: a SetSystem
+        """
+        cdef SetSystem NSC
+        NSC = SetSystem(list(self.groundset()))
+        for i in self._k_C:
+            if i <= self.rank():
+                for C in self._k_C[i]:
+                    NSC.append(C)
+        return NSC
+
+    def no_broken_circuits_sets(self, order=None):
```

Comment:
This is an iterator, but the other methods (at least by a spot check) return lists. This should be a list for consistency (perhaps good to change to an iterator at some point though).

Reply:
Thanks, this needed some essential changes.